File: packages/agent/agent/tools/read_concept.ts
import { getDb } from "@clara-financas/db";
import { concepts, BUNDLES } from "@clara-financas/db/schema/knowledge";
import { forTenant } from "@clara-financas/db/tenant-scope";
import { and, eq, like } from "drizzle-orm";
import { defineTool } from "eve/tools";
import { z } from "zod";

import { optionalText } from "../lib/schema";
import { requireTenantCaller } from "../lib/tenant";

/**
 * Lê conceitos OKF do espaço do usuário.
 *
 * Note o que NÃO está no inputSchema: `tenantId`. Ele vem de
 * `requireTenantCaller(ctx)`, derivado da sessão autenticada. Um tenantId
 * aceito por parâmetro seria bug de segurança, não parâmetro — o modelo não
 * decide de quem é o dado.
 *
 * Toda leitura passa por `forTenant`, então a RLS do banco vale mesmo que algo
 * acima falhe.
 */
export default defineTool({
  description:
    "Lê conceitos do conhecimento do usuário. Use para consultar categorias, convenções e regras da constituição, ou o que já foi aprendido sobre a pessoa, antes de decidir categoria ou aplicar regra.",
  inputSchema: z.object({
    bundle: z
      .enum(BUNDLES)
      .describe(
        "constitution para o contrato do domínio (categorias, convenções, regras); learnings para o que foi aprendido sobre esta pessoa.",
      ),
    // optionalText trata "" como ausente: modelos preenchem campos opcionais
    // com string vazia, e isso viraria filtro `= ''` — que não casa com nada.
    conceptId: optionalText().describe(
      'ID exato do conceito, ex.: "categories/groceries". Omita para listar o bundle inteiro.',
    ),
    type: optionalText().describe(
      'Filtra por tipo, ex.: "Category", "AlertRule", "Convention".',
    ),
    prefix: optionalText().describe('Filtra por prefixo de caminho, ex.: "categories/".'),
  }),
  async execute(input, ctx) {
    const { tenantId } = requireTenantCaller(ctx);

    return forTenant(tenantId, async (tx) => {
      const filters = [eq(concepts.bundle, input.bundle)];

      if (input.conceptId !== undefined) {
        filters.push(eq(concepts.conceptId, input.conceptId));
      }
      if (input.type !== undefined) {
        filters.push(eq(concepts.type, input.type));
      }
      if (input.prefix !== undefined) {
        filters.push(like(concepts.conceptId, `${input.prefix}%`));
      }

      const rows = await tx
        .select({
          conceptId: concepts.conceptId,
          type: concepts.type,
          frontmatter: concepts.frontmatter,
          body: concepts.body,
        })
        .from(concepts)
        .where(and(...filters))
        .limit(50);

      // Numa listagem o corpo inteiro é ruído; num conceito específico é a
      // resposta. Enviar tudo sempre gastaria contexto à toa.
      const isSingle = input.conceptId !== undefined;

      return {
        bundle: input.bundle,
        count: rows.length,
        concepts: rows.map((row) => ({
          id: row.conceptId,
          type: row.type,
          title: row.frontmatter.title,
          description: row.frontmatter.description,
          ...(isSingle ? { body: row.body } : {}),
        })),
      };
    }, getDb());
  },
});

File: services/agent/clara/tools/read_concept.py
"""Lê conceitos OKF do espaço do usuário — porta de `agent/tools/read_concept.ts`.

Leitura pura, sem gate: consulta categorias, convenções e regras da
constituição, ou o que já foi aprendido sobre a pessoa.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from sqlalchemy import select
from sqlalchemy.orm import Session

from clara.db.models import Bundle, Concept

# O que passa disso é anunciado, nunca cortado calado.
LIMIT = 50


@dataclass(frozen=True)
class ConceptSummary:
    id: str
    type: str
    title: str | None
    description: str | None
    body: str | None = None


@dataclass(frozen=True)
class ReadConceptResult:
    bundle: Bundle
    count: int
    concepts: list[ConceptSummary] = field(default_factory=list)
    truncated: bool = False
    note: str | None = None


def read_concept(
    session: Session,
    tenant_id: str,
    *,
    bundle: Bundle,
    concept_id: str | None = None,
    concept_type: str | None = None,
    prefix: str | None = None,
) -> ReadConceptResult:
    conditions = [Concept.tenant_id == tenant_id, Concept.bundle == bundle]
    if concept_id is not None:
        conditions.append(Concept.concept_id == concept_id)
    if concept_type is not None:
        conditions.append(Concept.type == concept_type)
    if prefix is not None:
        conditions.append(Concept.concept_id.like(f"{prefix}%"))

    # Pede um a mais que o teto só para saber se havia mais.
    rows = (
        session.execute(
            select(Concept).where(*conditions).order_by(Concept.concept_id).limit(LIMIT + 1)
        )
        .scalars()
        .all()
    )

    truncated = len(rows) > LIMIT
    page = rows[:LIMIT] if truncated else rows
    # Numa listagem o corpo inteiro é ruído; num conceito específico é a
    # resposta.
    is_single = concept_id is not None

    return ReadConceptResult(
        bundle=bundle,
        count=len(page),
        truncated=truncated,
        note=(
            f"Mostrando {LIMIT} conceitos; há mais. Filtre por concept_type ou prefix para ver "
            "o resto."
            if truncated
            else None
        ),
        concepts=[
            ConceptSummary(
                id=row.concept_id,
                type=row.type,
                title=row.frontmatter.get("title"),
                description=row.frontmatter.get("description"),
                body=row.body if is_single else None,
            )
            for row in page
        ],
    )
